icu4c/source/common/wintz.cpp

```diff
@@ -213,184 +53,26 @@ static LONG getTZKeyName(char* tzKeyName, int32_t tzKeyNamelength)
 U_CFUNC const char* U_EXPORT2
 uprv_detectWindowsTimeZone() 
 {
-    UErrorCode status = U_ZERO_ERROR;
-    UResourceBundle* bundle = NULL;
     char* icuid = NULL;
-    char apiStdName[MAX_LENGTH_ID];
-    char regStdName[MAX_LENGTH_ID];
-    char tmpid[MAX_LENGTH_ID];
-    int32_t len;
-    int id;
-    int errorCode;
-    wchar_t ISOcodeW[3]; /* 2 letter iso code in UTF-16*/
-    char  ISOcodeA[3]; /* 2 letter iso code in ansi */
-
-    LONG result;
-    TZI tziKey;
-    TZI tziReg;
-    TIME_ZONE_INFORMATION apiTZI;
-
-    BOOL tryPreVistaFallback;
-    OSVERSIONINFO osVerInfo;
-
-    /* Obtain TIME_ZONE_INFORMATION from the API, and then convert it
-       to TZI.  We could also interrogate the registry directly; we do
-       this below if needed. */
-    uprv_memset(&apiTZI, 0, sizeof(apiTZI));
-    uprv_memset(&tziKey, 0, sizeof(tziKey));
-    uprv_memset(&tziReg, 0, sizeof(tziReg));
-    GetTimeZoneInformation(&apiTZI);
-    tziKey.bias = apiTZI.Bias;
-    uprv_memcpy((char *)&tziKey.standardDate, (char*)&apiTZI.StandardDate,
-           sizeof(apiTZI.StandardDate));
-    uprv_memcpy((char *)&tziKey.daylightDate, (char*)&apiTZI.DaylightDate,
-           sizeof(apiTZI.DaylightDate));
-
-    /* Convert the wchar_t* standard name to char* */
-    uprv_memset(apiStdName, 0, sizeof(apiStdName));
-    wcstombs(apiStdName, apiTZI.StandardName, MAX_LENGTH_ID);
-
-    tmpid[0] = 0;
+    size_t len;
+    DYNAMIC_TIME_ZONE_INFORMATION dynamicTZI;
 
-    id = GetUserGeoID(GEOCLASS_NATION);
-    errorCode = GetGeoInfoW(id, GEO_ISO2, ISOcodeW, 3, 0);
-    u_strToUTF8(ISOcodeA, 3, NULL, (const UChar *)ISOcodeW, 3, &status);
+    /* Obtain DYNAMIC_TIME_ZONE_INFORMATION from the API */
+    uprv_memset(&dynamicTZI, 0, sizeof(dynamicTZI));
+    GetDynamicTimeZoneInformation(&dynamicTZI);
 
-    bundle = ures_openDirect(NULL, "windowsZones", &status);
-    ures_getByKey(bundle, "mapTimezones", bundle, &status);
-
-    /*
-        Windows Vista+ provides us with a "TimeZoneKeyName" that is not localized
-        and can be used to directly map a name in our bundle. Try to use that first
-        if we're on Vista or higher
-    */
-    uprv_memset(&osVerInfo, 0, sizeof(osVerInfo));
-    osVerInfo.dwOSVersionInfoSize = sizeof(osVerInfo);
-    tryPreVistaFallback = TRUE;
-    result = getTZKeyName(regStdName, sizeof(regStdName));
-    if(ERROR_SUCCESS == result) 
+    /* Copy the timezone ID to icuid to be returned. */
+    if (dynamicTZI.TimeZoneKeyName[0] != 0)
     {
-        UResourceBundle* winTZ = ures_getByKey(bundle, regStdName, NULL, &status);
```

Comment:
I think you still need to look up the ICU time zone in the resources. You can't just take the Windows time zone string from the API and just return that.
You need to handle the Windows -> IANA/CLDR time zone mapping.